library/net.cpp

```diff
@@ -63,6 +65,94 @@ enum dnet_socket_state {
  * @ok will be set to 1 if given socket was successfully initialized (connected or made listened)
  */
 struct dnet_addr_socket {
+	dnet_addr_socket()
+	: s(-1),
+	 buffer(nullptr)
+	{}
+
+	~dnet_addr_socket() {
+		if (s >= 0)
+			close(s);
+
+		free(buffer);
+	}
+
+	static int create_socket(dnet_node *node, const dnet_addr *address, int listening) {
+		socklen_t salen;
+		sockaddr *sa;
+		dnet_net_state *st;
+		int s;
+		int err;
+
+		st = dnet_state_search_by_addr(node, address);
+		if (st) {
+			err = -EEXIST;
+
+			dnet_log(node, DNET_LOG_NOTICE, "Address %s already exists in route table",
+				 dnet_addr_string(address));
+			dnet_state_put(st);
+			return err;
+		}
+
+		salen = address->addr_len;
+		sa = (sockaddr *)address;
+
+		sa->sa_family = address->family;
+
+		s = socket(sa->sa_family, SOCK_STREAM, IPPROTO_TCP);
+		if (s < 0) {
+			err = -errno;
+
+			dnet_log_err(node, "Failed to create socket for %s: family: %d",
+				     dnet_addr_string(address), sa->sa_family);
+			return err;
+		}
+
+		fcntl(s, F_SETFL, O_NONBLOCK);
+		fcntl(s, F_SETFD, FD_CLOEXEC);
+
+		if (listening) {
+			err = 1;
+			setsockopt(s, SOL_SOCKET, SO_REUSEADDR, &err, 4);
+
+			err = bind(s, sa, salen);
+			if (err) {
+				err = -errno;
+				dnet_log_err(node, "Failed to bind to %s",
+					     dnet_addr_string(address));
+				return err;
+			}
+
+			err = listen(s, 10240);
+			if (err) {
+				err = -errno;
+				dnet_log_err(node, "Failed to listen at %s",
+					     dnet_addr_string(address));
+				return err;
+			}
+
+			dnet_log(node, DNET_LOG_INFO, "Server is now listening at %s.",
+				 dnet_addr_string(address));
+		} else {
+			dnet_log(node, DNET_LOG_INFO, "Added %s to connect list",
+				 dnet_addr_string(address));
+		}
+
+		return s;
+	}
+
+	int init(dnet_node *node, const dnet_addr *address, int listening) {
+		s = create_socket(node, address, listening);
```

Comment:
init() may be successful or not, so in case of moving init() logic to c'tor causes necessity of throwing exception on errors. I think that elliptics shouldn't use exceptions mechanism for handling error cases.